### PR TITLE
Add preliminary implementation of find_tree_init_hash

### LINKED ISSUE #24: Add iterator for first-parent-iteration of commits
For some of the functionality we want to provide we need to trace commits back to an issue's initial message via the chain of commits' first-parents. An iterator which follows the first parents would be advantageous.
The iterator should be constructible from a `git2::Commit`, return the commit itself on the first call to `next()` and return the first-parent of the last commit returned in subsequent calls.

### PR DESCRIPTION
**Pull request by neithernut**
Targets #10. Resolves #24.

### CODE REVIEW

**Comment by neithernut** [comment]
For now, I'm reusing the `git2::Error` type. I guess @matthiasbeyer would like to use a new one for `git-dit`.

**Comment by matthiasbeyer** [comment]
Yep. I'd add `error_chain` to the dependencies for this.

**Comment by neithernut** [comment]
Do you want to open a PR for introducing an error type or should I do it in this one?

**Comment by matthiasbeyer** [comment]
I'm about to open one. Hold on.

**Comment by neithernut** [comment]
I will now rebase to get the clap setup.

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            description("TODO: Wrapped error")
+            display("TODO: Wrapped error")
+        }
+
+        SubCommandNotFound(subcommand: String) {
+            description("Unknown subcommand")
+            display("Unknown subcommand: {}", subcommand)
+        }
+    }
+}
```
Comment (neithernut) [inline src/error.rs]: This file was largely copied from `lib/src/error.rs`, which is licensed under the MPL. Are you ok with this, @matthiasbeyer?

**Comment by neithernut** [comment]
It doesn't work yet. It also lacks a check for the presence of the `commit` parameter.

**Comment by neithernut** [comment]
I also just realized that clap ships with an error type

File: src/main.rs [hunk 2]
```diff
@@ -9,12 +9,51 @@
...
+fn open_dit_repo() -> Result<Repository> {
+    // TODO: access the config and maybe return another repo instead
+    Repository::open_from_env().chain_err(|| EK::WrappedGitError)
+}
+
+
+/// find-tree-init-hash subcommand implementation
+///
+fn find_tree_init_hash(repo: &Repository, matches: &clap::ArgMatches) -> Result<()> {
+    repo.refname_to_id(matches.value_of("commit").unwrap())
```
Comment (neithernut) [inline src/main.rs]: Apparently, `refname_to_id()` doesn't do what I thought it does.

**Comment by neithernut** [comment]
Works now, apparently. Stll, I'm not sure whether I should discard the `FirstParentIter` in favour for [git2::Revwalk](https://docs.rs/git2/0.6.4/git2/struct.Revwalk.html) or if our use-case justifies the extra iterator (for convenience).

File: src/error.rs (hunk 1 above)
Comment (matthiasbeyer) [inline src/error.rs]: Yes.

**Comment by neithernut** [comment]
Blocks #28.

File: src/main.rs [hunk 3]
```diff
@@ -9,12 +9,51 @@
...
+fn open_dit_repo() -> Result<Repository> {
+    // TODO: access the config and maybe return another repo instead
+    Repository::open_from_env().chain_err(|| EK::WrappedGitError)
+}
+
+
+/// find-tree-init-hash subcommand implementation
+///
+fn find_tree_init_hash(repo: &Repository, matches: &clap::ArgMatches) -> Result<()> {
+    repo.revparse_single(matches.value_of("commit").unwrap())
```
Comment (matthiasbeyer) [inline src/main.rs]: Please add a short comment here that `unwrap()` is secured by clap.

File: src/main.rs [hunk 4]
```diff
@@ -9,12 +9,51 @@
...
     let matches = App::from_yaml(yaml).get_matches();
-    println!("Hello, world!");
+
+    let repo = open_dit_repo().unwrap();
+
+    match matches.subcommand() {
+        ("find-tree-init-hash", Some(sub_matches))  => find_tree_init_hash(&repo, sub_matches),
+        (name, _) => Err(Error::from_kind(EK::SubCommandNotFound(String::from(name)))),
+        // TODO: find and exec 3rd party executable
+    }.unwrap()
```
Comment (matthiasbeyer) [inline src/main.rs]: This panics on error (for example if `find_tree_init_hash()` returns an `Err(_)`.

What I'd do: Return nothing from `find_tree_init_hash()` and let the function itself do the printing. We could return an exit value from the function,... which then gets used in main() to `std::process::exit()`...

**Comment by matthiasbeyer** [comment]
> Works now, apparently. Stll, I'm not sure whether I should discard the FirstParentIter in favour for git2::Revwalk or if our use-case justifies the extra iterator (for convenience).

I'd use `git2::Revwalk` and wrap it in our iterator. Overhead is almost zero then, but we get the flexibility and we're able to tweak it later on without rewriting the codebase.

File: src/main.rs (hunk 4 above)
Comment (neithernut) [inline src/main.rs]: Yep. Originally, I planned on implementing proper error handling later on. But I'll implement the changes in this PR. This will also simplify implementing #28.

File: src/main.rs [hunk 5]
```diff
@@ -36,24 +36,26 @@ fn open_dit_repo() -> Result<Repository> {
...
+        .unwrap_or_else(|err| {error!("{}", err); 1})
 }
 
 
 fn main() {
     let yaml    = load_yaml!("cli.yaml");
     let matches = App::from_yaml(yaml).get_matches();
 
-    let repo = open_dit_repo().unwrap();
+    let repo = open_dit_repo().unwrap(); // TODO: error handling
```
Comment (matthiasbeyer) [inline src/main.rs]: `TODO: Fix panic` would be way more accurate :smile: 

Also, we can do a simple 

```
let repo = match open_dir_repo() { 
    Ok(r) => r, 
    Err(e) => { 
        error!("{}", e); 
        std::process::exit(1) 
    }
}
```

 here

File: src/main.rs [hunk 6]
```diff
@@ -36,24 +36,26 @@ fn open_dit_repo() -> Result<Repository> {
...
 
-    match matches.subcommand() {
+    std::process::exit(match matches.subcommand() {
         ("find-tree-init-hash", Some(sub_matches))  => find_tree_init_hash(&repo, sub_matches),
-        (name, _) => Err(Error::from_kind(EK::SubCommandNotFound(String::from(name)))),
```
Comment (matthiasbeyer) [inline src/main.rs]: Why did you remove this?

File: src/main.rs (hunk 5 above)
Comment (neithernut) [inline src/main.rs]: Well, yeah. I was kinda lazy.

File: src/main.rs (hunk 6 above)
Comment (neithernut) [inline src/main.rs]: I did not remote this but replace it with direct error printing.

Comment (matthiasbeyer) [inline src/main.rs]: Whops. sure! :laughing:

**Comment by neithernut** [comment]
About the `FirstParentIter`: apparently `Revwalk` must be pulled from a repo while our iterator can be created from a commit. Hence, I'd keep the current implementation.

File: src/main.rs (hunk 5 above)
Comment (neithernut) [inline src/main.rs]: I also did not realize that this doesn't generate any conflicts between the branches' return type.

**Comment by neithernut** [comment]
I'm ready for squashing.

**Comment by matthiasbeyer** [comment]
Go ahead.

**Comment by neithernut** [comment]
@matthiasbeyer can you give me an ACK/NACK on this?

File: lib/src/lib.rs [hunk 7]
```diff
@@ -8,7 +8,11 @@
 //
 
 #[macro_use] extern crate log;
+#[macro_use] extern crate error_chain;
 extern crate git2;
 
+pub mod error;
 pub mod repository;
 
+mod first_parent_iter;
```
Comment (matthiasbeyer) [inline lib/src/lib.rs]: I'm rather sure we need to make this `pub` or export the iterator in another way... But I guess we can do so later.
